sdk/core/core-client-rest/src/urlHelpers.ts

```diff
@@ -60,6 +34,44 @@ export function buildRequestUrl(
   );
 }
 
+function appendQueryParams(url: string, options: RequestParameters = {}) {
+  if (!options.queryParameters) {
+    return url;
+  }
+  let parsedUrl = new URL(url);
+  const queryParams = options.queryParameters;
+  for (const key of Object.keys(queryParams)) {
+    const param = queryParams[key] as any;
+    if (param === undefined || param === null) {
+      continue;
+    }
+    if (!param.toString || typeof param.toString !== "function") {
+      throw new Error(`Query parameters must be able to be represented as string, ${key} can't`);
+    }
+    const value = param.toISOString !== undefined ? param.toISOString() : param.toString();
+    parsedUrl.searchParams.append(key, value);
+  }
+
+  if (options.skipUrlEncoding) {
+    parsedUrl = skipQueryParameterEncoding(parsedUrl);
+  }
+  return parsedUrl.toString();
+}
+
+function skipQueryParameterEncoding(url: URL) {
```

Comment:
maybe a small comment about how this function helps with avoiding how `URLSearchParams` performs `application/x-www-form-urlencoded` on inputs, which causes issues on the server side?

Reply:
The URL encoding is NOT an idempotent operation.  For example, the percent character "%" is URL-encoded as "%25". If the parameter is URL-encoded again, "%25" becomes "%2525". 

So this happens if the query value already encoded in client code so setting `skipUrlEncoding: true` could ensure parameter only encoded once.

Reply:
> maybe a small comment about how this function helps with avoiding how `URLSearchParams` performs `application/x-www-form-urlencoded` on inputs, which causes issues on the server side?

The URL encoding is NOT an idempotent operation. For example, the percent character "%" is URL-encoded as "%25". If the parameter is URL-encoded again, "%25" becomes "%2525".

So this happens if the query value already encoded in client code so setting skipUrlEncoding: true could ensure parameter only encoded once.

Reply:
So this flag is to opt-out of core-client-rest doing the encoding, since the client has already taken care of encoding it? Just to ask the question, why not have the client skip encoding it and encode it here instead?

I think what I was trying to get at is URLSearchParams does *extra* encoding than are needed for query parameters, so avoiding that make sense to me, but we still need to in some place make sure we're sending a valid URL by having query parameters be well-formed

Reply:
I have to clarify that what I am talking about the client are two main users, 1) the end user to call our generated RLC and 2) the service team to provide the swagger file.  The former could enable the option `skipUrlEncoding = true` when calling method, the latter could enable the extension in swagger file by [the doc](https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-skip-url-encoding).

By default encoding is the good choice, but if the exception happens they could opt-out of core-client-rest doing the encoding. This is my understanding.